### DISCUSSION TITLE
DEF_TIM function

### DISCUSSION THREAD

**Discussion by kvng1** [iNavFlight/inav#6465] [Q&A]
So I've just built a custom firmware for my F411-WSE, while it seems working fine on the bench, I still don't understand what is the meaning of last 2 parameters, sometime they use 0,0 or 0,1. Can anyone give me some insight? Thanks.

    DEF_TIM(TIM3, CH1, PB4,  TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR,  0, 0), // S1  D(1,4,5)
    DEF_TIM(TIM3, CH2, PB5,  TIM_USE_MC_MOTOR | TIM_USE_FW_MOTOR,  0, 0), // S2  D(1,5,5)
    DEF_TIM(TIM1, CH1, PA8,  TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO,  0, 1), // S3  D(2,1,6)
    DEF_TIM(TIM1, CH2, PA9,  TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO,  0, 1), // S4  D(2,2,6)
    DEF_TIM(TIM1, CH3, PA10, TIM_USE_MC_MOTOR | TIM_USE_FW_SERVO,  0, 1), // S5  D(2,6,6)
    DEF_TIM(TIM4, CH3, PB8,  TIM_USE_MC_SERVO | TIM_USE_FW_SERVO,  0, 0), // S6  D(1,7,2)

    DEF_TIM(TIM2, CH2, PA1,  TIM_USE_MC_SERVO | TIM_USE_FW_SERVO,  0, 0), // Air Speed Pad
    DEF_TIM(TIM2, CH3, PB10, TIM_USE_MC_SERVO | TIM_USE_FW_SERVO,  0, 0), //LED 2812 Pad

    DEF_TIM(TIM11, CH1, PB9,  TIM_USE_ANY,   0, 0), //ST1 pad - softserial1_tx
    DEF_TIM(TIM5, CH1, PA0, TIM_USE_ANY,  0, 0), //RSSI pad - softserial1_rx

    DEF_TIM(TIM9, CH2, PA3,   TIM_USE_PPM,   0, 0), //RX2 Pad -PPM`

**Comment by damiendon** [comment]
Hi, have you an answer about that? I'm interested too !

**Reply by kvng1** [reply]
There is some info involving DMA channels and alternate function tables in the ardupilot document but I don't have time to look further at it yet.

**Reply by JulioCesarMatias** [reply]
0 = no;
1 = yes;

`#define DEF_TIM(tim, ch, pin, usage, flags, dmavar)` 

flags = use AF (Alternative function)? 
dmavar = use DMA with GPIO?

References:
Alternative Function -> https://community.st.com/s/question/0D50X00009Xkak4SAB/gpio-af-mode
DMA with GPIO -> https://community.st.com/s/question/0D50X0000C7eDobSQE/dma-gpio-stm32f103

**Reply by kedeng** [reply]
It took me some time to figure it out what "dmavar" mean.
See the picture below, "dmavar" is index of DEF_TIM_DMA__BTCH_TIM?_CH? which identify to use which Channel and Stream in DMA.

![image](https://github.com/user-attachments/assets/810fe490-453b-49c6-8d1f-c15adb13eba6)
